overlord/snapstate/target.go

```diff
@@ -0,0 +1,588 @@
+type Options struct {
+	// Flags contains flags that apply to the entire operation.
+	Flags Flags
+	// UserID is the ID of the user that is performing the operation.
+	UserID int
+	// DeviceCtx is an optional device context that will be used during the
+	// operation.
+	DeviceCtx DeviceContext
+	// PrereqTracker is an optional prereq tracker that will be used to keep
+	// track of all snaps (explicitly requested and implicitly required snaps)
+	// that might need to be installed during the operation.
+	PrereqTracker PrereqTracker
+	// FromChange is the change that triggered the operation.
+	FromChange string
+	// Seed should be true while seeding the device. This indicates that we
+	// shouldn't require that the device is seeded before installing/updating
+	// snaps.
+	Seed bool
+	// RequireOneSnap is a boolean flag indicating that this operation is expected
+	// to only operate on one snap (excluding any prerequisite snaps that may be
+	// required). If this is true, then the operation will fail if more than one
+	// snap is being operated on. This flag primarily exists to support the
+	// pre-existing behavior of calling InstallMany with one snap vs calling
+	// Install.
+	RequireOneSnap bool
+}
+
+// target represents the data needed to setup a snap for installation.
+type target struct {
+	// setup is a partially initialized SnapSetup that contains the data needed
+	// to find the snap file that will be installed.
+	setup *SnapSetup
+	// components is a list of partially initialized ComponentSetup structs that
+	// contain the data needed to find the component files that will be
+	// installed.
+	components []*ComponentSetup
+	// info contains the snap.info for the snap to be installed.
+	info *snap.Info
+	// snapst is the current state of the target snap, prior to installation.
+	// This must be retrieved prior to unlocking the state for any reason (for
+	// example, talking to the store).
+	snapst SnapState
+}
+
+// InstallGoal represents a single snap or a group of snaps to be installed.
+type InstallGoal interface {
+	// toInstall returns the data needed to setup the snaps for installation.
+	toInstall(context.Context, *state.State, Options) ([]target, error)
+}
+
+// storeInstallGoal implements the InstallGoal interface and represents a group of
+// snaps that are to be installed from the store.
+type storeInstallGoal struct {
+	snaps map[string]StoreSnap
+}
+
+// StoreSnap represents a snap that is to be installed from the store.
+type StoreSnap struct {
+	// InstanceName is the name of snap to install.
+	InstanceName string
+	// Components is the list of components to install with this snap.
+	Components []string
+	// RevOpts contains options that apply to the installation of this snap.
+	RevOpts RevisionOptions
+	// SkipIfPresent indicates that the snap should not be installed if it is already present.
+	SkipIfPresent bool
+}
+
+// StoreInstallGoal creates a new storeInstallGoal, which implements InstallGoal to
```

Comment:
we want to rephrase this leaving out the internal name, something like:

creates/returns an implementation of InstallGoal to install/installing ...